pathd/path_zebra.c

```diff
@@ -558,40 +578,79 @@ static void path_nht_update(struct prefix *nh, int nhnum)
 			continue;
 		if (!IPV6_ADDR_SAME(&sid_srv6.u.prefix6, &nh->u.prefix6))
 			continue;
-		if (nhnum)
-			path_zebra_add_sr_policy(policy, segment_list);
+		if (nhtd->nh_num)
+			path_zebra_add_sr_policy_internal(policy, segment_list,
+							  nhtd);
 		else
 			path_zebra_delete_sr_policy(policy);
 	}
 }
 
-static int path_zebra_nexthop_update(ZAPI_CALLBACK_ARGS)
+static bool path_zebra_srv6_nexthop_info_update(struct path_nht_data *nhtd,
+						struct zapi_route *nhr)
 {
-	struct path_nht_data *nhtd, lookup;
-	struct zapi_route nhr;
-	struct prefix matched;
+	struct nexthop *nexthop;
+	struct nexthop *nhlist_head = NULL;
+	struct nexthop *nhlist_tail = NULL;
+	struct nexthop *oldnh;
+	bool nh_changed = false;
+	int i;
+
+	if (nhtd && nhr)
+		nhtd->nh_num = nhr->nexthop_num;
+
+	if (!nhr->nexthop_num) {
+		nhtd->nh_num = nhr->nexthop_num;
+		if (nhtd->nexthop)
+			nexthop_free(nhtd->nexthop);
+		nhtd->nexthop = NULL;
+		return true;
+	}
 
-	if (!zapi_nexthop_update_decode(zclient->ibuf, &matched, &nhr)) {
-		zlog_err("Failure to decode nexthop update message");
-		return 1;
+	for (i = 0; i < nhr->nexthop_num; i++) {
+		nexthop = nexthop_from_zapi_nexthop(&nhr->nexthops[i]);
+
+		if (nhlist_tail) {
+			nhlist_tail->next = nexthop;
+			nhlist_tail = nexthop;
+		} else {
+			nhlist_tail = nexthop;
+			nhlist_head = nexthop;
+		}
+
+		for (oldnh = nhtd->nexthop; oldnh; oldnh = oldnh->next)
+			if (nexthop_same(oldnh, nexthop))
+				break;
+
+		if (!oldnh)
+			nh_changed = true;
 	}
-	if (matched.family != AF_INET6)
-		return 1;
+	if (nhtd->nexthop)
+		nexthop_free(nhtd->nexthop);
+	nhtd->nexthop = nhlist_head;
+	nhr->nexthop_num = nhr->nexthop_num;
+
+	return nh_changed;
+}
+
+static void path_zebra_nexthop_update(struct vrf *vrf, struct prefix *match,
+				      struct zapi_route *nhr)
+{
+	struct path_nht_data *nhtd, lookup;
+
+	if (match->family != AF_INET6)
+		return;
 
 	memset(&lookup, 0, sizeof(lookup));
-	lookup.nh = matched;
-	lookup.nh_vrf_id = vrf_id;
+	prefix_copy(&lookup.nh, match);
+	lookup.nh_vrf_id = vrf->vrf_id;
```

Comment:
to make consistency with each commit, the code changes in path_zebra_nexthop_update() should be moved to the previous "pathd: add registration mechanism"

Reply:
fixed